### PR TITLE
Add pubsub and mq implementation

### PR DESCRIPTION
**Pull request by madcake**

### CODE REVIEW

**Review by hewigovens — APPROVED** [review]
No existing pub sub package to use？

**Comment by madcake** [comment]
> No existing pub sub package to use？

No found for us

File: pubsub/mqclient/mq_stream.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+}
+
+func (s *Stream) delivery(msg amqp.Delivery) {
+	if (*s.consumer).Callback(msg) == nil {
+		ack((*s.consumer).GetQueue(), msg)
+	}
+}
+
+func ack(queue string, msg amqp.Delivery) {
```
Comment (vikmeup) [inline pubsub/mqclient/mq_stream.go:90]: should this return an error?

File: pubsub/mqclient/mq_client.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-time.After(reconnectDelay + time.Second):
+				// Add metric
+			}
+		}
+		select {
+		case <-c.ctx.Done():
+			log.Fatal("Fucking end")
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:177]: Fix log

File: pubsub/mqclient/mq_client.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	c.notifyClose = make(chan *amqp.Error)
+	c.notifyConfirm = make(chan amqp.Confirmation)
+	c.channel.NotifyClose(c.notifyClose)
+	c.channel.NotifyPublish(c.notifyConfirm)
+	c.isConnected.Store(true)
+	return nil
+}
+
+func (c *Client) Run() error {
+	log.Debug("Interface: ", c)
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:70]: any reason to log interface?

File: pubsub/mqclient/mq_client.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+			ContentType: "text/plain",
+			Body:        data,
+		},
+	)
+}
+
+func (c *Client) Close() error {
+	err := c.channel.Close()
+	if err != nil {
+		return errors.New("Client.Close MQ issue " + err.Error())
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:149]: Can you add client close channel?

File: pubsub/mqclient/mq_client.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	err = c.conn.Close()
+	if err != nil {
+		return errors.New("Client.Close MQ issue " + err.Error())
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:154]: Can you add client connection channel?

File: pubsub/mqclient/mq_client.go (hunk 4 above)
Comment (vikmeup) [reply pubsub/mqclient/mq_client.go:149]: because logs are the same for both at the moment

File: pubsub/mqclient/mq_stream.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func (s *Stream) IsWriteOnly() bool {
+	return s.isWriteOnly
+}
+
+func (s *Stream) declareQueue() {
+	_, err := s.channel.QueueDeclare((*s.consumer).GetQueue(), true, false, false, false, nil)
+	if err != nil {
+		log.Fatal("Stream.Init MQ issue "+err.Error(), (*s.consumer).GetQueue())
```
Comment (vikmeup) [inline pubsub/mqclient/mq_stream.go:80]: wrong log? this is `declareQueue`

File: pubsub/mqclient/mq_client.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	ch, err := conn.Channel()
+	if err != nil {
+		log.Error("Client.connect MQ Channel issue " + err.Error())
+		return err
+	}
+	err = ch.Confirm(false)
+	if err != nil {
+		log.Error("Client.connect MQ Channel issue " + err.Error())
+	}
+	log.Debug("Interface: ", c)
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:58]: any reason for this log?

File: pubsub/mqclient/mq_client.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+func (c *Client) Connect() error {
+	conn, err := amqp.Dial(c.uri)
+	if err != nil {
+		log.Error("Client.connect MQ Dial issue " + err.Error())
+		return err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		log.Error("Client.connect MQ Channel issue " + err.Error())
```
Comment (vikmeup) [inline pubsub/mqclient/mq_client.go:51]: same log as below: `log.Error("Client.connect MQ Channel issue " + err.Error())`. Can you give unique names?

File: pubsub/mqclient/mq_stream.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+func (s *Stream) IsConnected() bool {
+	return s.isConnected.Load()
+}
+
+func (s *Stream) IsWriteOnly() bool {
+	return s.isWriteOnly
+}
+
+func (s *Stream) declareQueue() {
```
Comment (vikmeup) [inline pubsub/mqclient/mq_stream.go:77]: should this return an error?

File: tests/integration/pubsub/pubsub_mq_test.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		log.Fatalf("Could not connect to docker: %s", err)
+	}
+
+	mqResource, err = pool.Run("rabbitmq", "latest", nil)
+	if err != nil {
+		log.Fatalf("Could not start resource: %s", err)
+	}
+	//time.Sleep(10 * time.Second)
```
Comment (vikmeup) [inline tests/integration/pubsub/pubsub_mq_test.go:78]: remove log

File: tests/integration/pubsub/pubsub_mq_test.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	mqClient   pubsub.Client
+	ctx        context.Context
+)
+
+type RawTransactionConsumer struct {
+	TestFunc func(delivery amqp.Delivery)
+}
+
+func (c RawTransactionConsumer) GetQueue() string {
+	return string("test_queue")
```
Comment (vikmeup) [inline tests/integration/pubsub/pubsub_mq_test.go:31]: Just `"test_queue"`

**Pull request merged**